First version of `libqlever` for an embedded use of QLever

joka921 opened the pull request:
This enables writing C++ code that builds an index for a given dataset, loads such an index from disk, and issue queries in an embedded fashion, that is, without running a HTTP server. For a usage example, have a look at `src/libqlever/LibQLeverExample.cpp`. This can be compiled with `ninja -C build LibQLeverExample`, which produces a binary `LibQLeverExample` in the `build` folder. To compile other code in this vein, just add the corresponding lines to `CMakeLists.txt` in the top-level folder (search for `LibQLeverExample`). Some notes and limitations:

1. This is a first minimal but functional and useful version. For an overview of what can be configured already now, see `src/libqlever/Qlever.h`
2. To get detailed timings on the various parts of the query execution (like when you click "Analysis" in the QLever UI), use `MediaType::qleverJson`
3. To compile code like that in `LibQLeverExample.cpp`, QLever's HTTP server code needs to be neither compiled nor linked
4. The code for `IndexBuilderMain` already uses `libqlever` now and has become much simpler
5. Some of the current limitations: timeouts, manual cancellation, live timings while the query is running, updates. These are all supported by QLever in the standard client/server setup, but not yet by `libqlever`

joka921 committed:
Fix libqlever.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

joka921 committed:
Refactor the libqlever.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

joka921 committed:
Try and fix the compilation.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

joka921 committed:
Merge branch 'master' into libqlever-clean-for-merge

# Conflicts:
#	src/index/IndexBuilderMain.cpp

joka921 committed:
Fix this.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

joka921 committed:
Reenable the logging, to debug the conformance issues.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

ullingerc commented on src/libqlever/Qlever.h:
```diff
@@ -0,0 +1,134 @@
+  // index will be built.
+  std::vector<qlever::InputFileSpecification> inputFiles;
+
+  std::optional<ad_utility::MemorySize> parserBufferSize;
```
Add an explanation here

ullingerc commented on src/libqlever/Qlever.h:
```diff
@@ -0,0 +1,134 @@
+  // TODO<joka921> Document these additional settings.
+  std::string settingsFile;
+
+  ad_utility::VocabularyType vocabType{
```
And also here

ullingerc commented on src/libqlever/Qlever.h:
```diff
@@ -0,0 +1,134 @@
+  explicit EngineConfig(const IndexBuilderConfig& c)
+      : CommonConfig{static_cast<const CommonConfig&>(c)} {}
+  EngineConfig() = default;
+  bool text = false;
+  bool persistUpdates = true;
```
Also comment these, especially the first one, which doesn't have a very telling name

ullingerc commented on src/engine/CMakeLists.txt:
```diff
@@ -18,3 +18,6 @@ add_library(engine
         NeutralOptional.cpp Load.cpp)
 qlever_target_link_libraries(engine util index parser sparqlExpressions http SortPerformanceEstimator Boost::iostreams s2 spatialjoin-dev pb_util)
+
+add_library(server Server.cpp)
```
I would have assumed the server needs to be linked against the other libraries, but as the checks run through, it seems this is fine

joka921 replied:
It is probably cleaner to linkt them here explicitly.

ullingerc commented on src/ServerMain.cpp:
```diff
@@ -164,7 +158,7 @@ int main(int argc, char** argv) {
   try {
     Server server(port, numSimultaneousQueries, memoryMaxSize,
-                  std::move(accessToken), !noPatternTrick);
+                  std::move(accessToken), !noPatterns);
```
Can you briefly explain why you remove this option and how it is equivalent to do what you do now? I'm not familiar with this.

joka921 replied:
As the socalled `PatternTrick` is the only usage of the `pattern` feature (without  going into detail here what it is), having to separate options here seemed rather redundant, and in practice they were never set to different values.

ullingerc commented on src/libqlever/Qlever.cpp:
```diff
@@ -0,0 +1,164 @@
+
+// ___________________________________________________________________________
+std::string Qlever::query(std::string query, ad_utility::MediaType mediaType) {
+  return this->query(parseAndPlanQuery(std::move(query)), mediaType);
```
Is there a particular reason to use `this->` here?

ullingerc reviewed (changes requested):
Some small remarks and a question. Please also address the sonarcloud comments on const correctness.

joka921 committed:
Merge branch 'master' into libqlever-clean-for-merge

# Conflicts:
#	src/engine/CMakeLists.txt
#	src/index/IndexBuilderMain.cpp

joka921 committed:
Add some unit tests and make sonarcloud happy etc.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

ullingerc reviewed (commented):
In addition to Robin's comments, I only have one question

joka921 commented on src/libqlever/Qlever.cpp:
```diff
@@ -0,0 +1,166 @@
+    LOG(ERROR) << "Creating the index for QLever failed with the following "
+                  "exception: "
+               << e.what() << std::endl;
+    throw;
```
Moved this logic into the respective main functions.

joka921 commented on src/libqlever/Qlever.cpp:
```diff
@@ -0,0 +1,166 @@
+  const auto& [qet, qec, parsedQuery] = queryPlan;
+  ad_utility::Timer timer{ad_utility::Timer::Started};
+  // TODO<joka921> For cancellation we have to call
+  // `recursivelySetCancellationHandle` (see `Server::parseAndPlan`).
```
I've added a note to the docstring of the function.

joka921 committed:
Fix the compilation

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

joka921 committed:
Add unit tests.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

joka921 committed:
Address some sonarcloud issues.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

ullingerc reviewed (approved):
As soon as the queued checks pass, this should be good to go imo

RobinTF commented on test/libqlever/QleverTest.cpp:
```diff
@@ -12,6 +12,118 @@
 
+// _____________________________________________________________________________
+TEST(LibQlever, buildIndexAndRunQuery) {
+  std::string filename = "testIndexForLibQlever.ttl";
```
I recently started calling test files like this just like their test name (`buildIndexAndRunQuery.ttl` in this case) to make name clashes unlikely. We should really consider creating a gtest helper function that generates temporary files for stuff just like this. The "current test case" is information that is available in a static context.

Something like

```cpp

inline std::string GetTempFilenameForCurrentTest(const std::string& extension = ".tmp") {
    static std::atomic<uint64_t> counter{0};

    const ::testing::TestInfo* test_info =
        ::testing::UnitTest::GetInstance()->current_test_info();
    if (!test_info) {
        throw std::runtime_error("No current test info available");
    }

    std::string suite = test_info->test_suite_name();
    std::string name  = test_info->name();

    // Sanitize for filesystem safety
    auto sanitize = [](std::string s) {
        for (char& c : s) {
            if (!std::isalnum(static_cast<unsigned char>(c)) && c != '_') {
                c = '_';
            }
        }
        return s;
    };

    suite = sanitize(suite);
    name  = sanitize(name);

    // Add timestamp + counter for uniqueness
    auto now = std::chrono::steady_clock::now().time_since_epoch();
    auto nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();

    uint64_t id = counter++;

    std::ostringstream oss;
    oss << suite << "_" << name << "_"
        << nanos << "_" << id << extension;

    std::filesystem::path temp_dir = std::filesystem::temp_directory_path();
    std::filesystem::path filename = temp_dir / oss.str();

    return filename.string();
}
```
Should already do the trick. (Untested AI-generated yadayada.)

joka921 replied:
yes at some point we can do this, I agree...

RobinTF commented on test/libqlever/QleverTest.cpp:
```diff
@@ -12,6 +12,118 @@
+
+  // Test the activation of the memory limit
+  c.memoryLimit_ = ad_utility::MemorySize::bytes(0);
+  EXPECT_ANY_THROW(Qlever::buildIndex(c));
```
Please either add a comment why any_throw is the correct assertion here, or expect a fixed exception type. Otherwise this function might throw for a wrong reason without this test failing. Same for the other ANY_THROW in this file)

RobinTF commented on test/libqlever/QleverTest.cpp:
```diff
@@ -12,6 +12,118 @@
+                            ad_utility::MediaType::tsv);
+    EXPECT_EQ(res, "?s\n<s>\n");
+
+    // Run with a different media type
```
```suggestion
    // Run with a different media type.
```

RobinTF commented on test/libqlever/QleverTest.cpp:
```diff
@@ -12,6 +12,118 @@
+    auto words = ad_utility::makeOfstream(wordsfileName);
+    words << "kartoffel\t0\t13\t1\n<s>\t1\t13\t1\n";
+    auto docs = ad_utility::makeOfstream(docsFileName);
+    docs << "13\tKartoffeln sind ein schönes Gemüse";
```
To be honest, for robustness you could append `🥔` at the end to make sure non-ascii characters are not resulting in exceptions.

Also there's `🥗` for the turtle file for the fun of it.

RobinTF reviewed (changes requested):
Only very minor requests and suggestions left, looking very good.

joka921 commented on src/libqlever/Qlever.h:
```diff
@@ -168,13 +171,15 @@ class Qlever {
   // Run the given query on the index. Currently only queries, but no updates
   // are supported.
+  // Note: Currently this function neither sets a timeout nor does support
+  // manual cancellation.
```
Yes but the comment remains valid: Using this interface you can not cancel the query for now.

joka921 committed:
Small changes from Robins review

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

RobinTF reviewed (approved):
Thank you very much.

Commit:
Merge remote-tracking branch 'origin/master' into libqlever-clean-for-merge

Commit:
Pass by Hannah

1. Revised comments
2. Make sure binary is in `build` and not `build/src/libqlever`
3. Improve `LibQLeverExample.cpp` (usability + documentation)

Commit:
Make Codespell happy
